File: Cpp/Game_of_survival/hardly_revivable_cell.cpp
#include "hardly_revivable_cell.h"

Model::HardlyRevivableCell::HardlyRevivableCell(bool is_alive) :
	is_alive_now(is_alive),
	CellBase(this->TYPE_CHAR)
{
}

const std::string& Model::HardlyRevivableCell::get_alive_state_sign() {
	static std::string obj = "\033[1;31;32m*\033[0m";
	return obj;
}

const std::string& Model::HardlyRevivableCell::get_dead_state_sign() {
	static std::string obj = "\033[1;31;31m*\033[0m";
	return obj;
}

bool Model::HardlyRevivableCell::is_perceived_as_alive() {
	return this->is_alive_now;
}

void Model::HardlyRevivableCell::stash_next_state()
{
	int alive_neighbours_count = this->count_alive_neighbours();
	bool will_be_alive;

	if (this->is_alive_now || this->iterations_to_wait_now > 0) {
		will_be_alive = alive_neighbours_count == 2 || alive_neighbours_count == 3;

		if (will_be_alive)
			this->iterations_to_wait_next = this->iterations_to_wait_now == 0 ?
			0 : this->iterations_to_wait_now - 1;
		else
			this->iterations_to_wait_next = 2;
	}
	else
		this->iterations_to_wait_next = 2;

	this->is_alive_next = this->iterations_to_wait_next == 0;
}

void Model::HardlyRevivableCell::release_next_state() {
	this->is_alive_now = this->is_alive_next;
	this->iterations_to_wait_now = this->iterations_to_wait_next;
}

const std::string& Model::HardlyRevivableCell::get_current_state_sign()
{
	return this->is_alive_now ?
		this->get_alive_state_sign() :
		this->get_dead_state_sign();
}

File: Kotlin/src/kotlin/game_of_survival/core/models/cells/HardlyRevivableCell.kt
package game_of_survival.core.models.cells

import game_of_survival.core.Constants

/**
 * Create a cell, that becomes alive after a prolonged period.
 * @param isAliveNow Is the cell alive at the moment.
 */
class HardlyRevivableCell(private var isAliveNow: Boolean) : CellBase() {
    private var iterationsToWaitNow = 0
    private var iterationsToWaitNext = 0
    private var isAliveNext: Boolean = false
    override val typeChar: Char = Constants.CellTypeChars.HARDLY_REVIVABLE_CELL
    override fun isPerceivedAsAlive(): Boolean = this.isAliveNow
    override fun stashNextState() {
        if (this.isAliveNow || this.iterationsToWaitNow > 0) {
            var aliveNeighboursCount = this.countAliveNeighbours()
            val willBeAlive = aliveNeighboursCount == 2 || aliveNeighboursCount == 3
            if (willBeAlive) {
                this.iterationsToWaitNext = if (this.iterationsToWaitNow == 0) {
                    0
                } else {
                    this.iterationsToWaitNow - 1
                }
            }
            else {
                this.iterationsToWaitNext = 2;
            }
        }
        else {
            this.iterationsToWaitNext = 2;
        }
        this.isAliveNext = this.iterationsToWaitNext == 0
    }
    override fun releaseNextState() {
        this.isAliveNow = this.isAliveNext
        this.iterationsToWaitNow = this.iterationsToWaitNext
    }
    override fun getCurrentStateSigning(): String =
        if (this.isAliveNow) {
            Constants.CellStateSignings.HardlyRevivable.ALIVE
        } else {
            Constants.CellStateSignings.HardlyRevivable.DEAD
        }
}
